src/base/DiagnosticPositionInfo.zig

```diff
@@ -0,0 +1,178 @@
+//! This module provides helpers for calculating position for diagnostics
+//! including the start and end line and column information
+const std = @import("std");
+const Allocator = std.mem.Allocator;
+
+start_line: u32,
+start_col: u32,
+end_line: u32,
+end_col: u32,
+line_text: []const u8,
+
+const DiagnosticPositionInfo = @This();
+
+/// Finds the line number for a given position in the source
+fn lineNum(newlines: std.ArrayList(u32), pos: u32) u32 {
+    if (newlines.items.len == 0) {
+        return 0;
+    }
+
+    var lineno: u32 = 0;
+
+    while (lineno + 1 < newlines.items.len) {
+        if (newlines.items[lineno + 1] > pos) {
+            return lineno;
+        }
+        lineno += 1;
+    }
+
+    return lineno;
+}
+
+/// Gets the column number for a position on a given line
+fn columnNum(newlines: std.ArrayList(u32), line: u32, pos: u32) !u32 {
+    const line_start: u32 = @intCast(newlines.items[line]);
+    if (pos < line_start) {
+        return error.InvalidPosition;
+    }
+    return pos - line_start;
+}
+
+/// Returns the source text for a given line
+fn getLineText(source: []const u8, newlines: std.ArrayList(u32), line: u32) []const u8 {
+    const line_start = newlines.items[line];
+    const line_end = if (line + 1 < newlines.items.len)
+        newlines.items[line + 1]
+    else
+        source.len;
+
+    return source[line_start..line_end];
+}
+
+/// Record the offsets for each newline in the source
+pub fn countNewlines(gpa: Allocator, source: []const u8) !std.ArrayList(u32) {
+    var newlines = std.ArrayList(u32).init(gpa);
+
+    try newlines.append(0);
+
+    // Find all newlines in the source
+    var pos: u32 = 0;
+    for (source) |c| {
+        if (c == '\n') {
+            try newlines.append(pos + 1); // Position after the newline
+        }
+        pos += 1;
+    }
+
+    return newlines;
+}
+
+/// Returns the position info for a diagnostic
+pub fn get(source: []const u8, newlines: std.ArrayList(u32), begin: u32, end: u32) !DiagnosticPositionInfo {
```

Comment:
100% optional, but might be nice to make a small struct that stores the source and newlines and can be use for this function. On init, it would count the newlines. Then it would be `diag_struct.get(begin, end)` at callsites.

Reply:
Thanks, I decided to keep it as is for now. I re-wired the toSExpr functions to take the line_starts ArrayList(u32) as an argument... but if we decide to do something fancier in future we can switch it to this.